lib/net/tcp_client/tcp_client.rb

```diff
@@ -221,19 +227,27 @@ def self.connect(params={})
     #   puts "Received: #{response}"
     #   client.close
     def initialize(parameters={})
-      params                  = parameters.dup
-      @read_timeout           = (params.delete(:read_timeout) || 60.0).to_f
-      @connect_timeout        = (params.delete(:connect_timeout) || (@read_timeout/2)).to_f
-      buffered                = params.delete(:buffered)
-      @buffered               = buffered.nil? ? true : buffered
-      @connect_retry_count    = params.delete(:connect_retry_count) || 10
-      @retry_count            = params.delete(:retry_count) || 3
-      @connect_retry_interval = (params.delete(:connect_retry_interval) || 0.5).to_f
-      @on_connect             = params.delete(:on_connect)
-      @server_selector        = params.delete(:server_selector) || :ordered
-      @close_on_error         = params.delete(:close_on_error)
-      @close_on_error         = true if @close_on_error.nil?
-      @logger                 = params.delete(:logger)
+      params                      = parameters.dup
+      @read_timeout               = (params.delete(:read_timeout) || 60.0).to_f
+      @write_timeout              = (params.delete(:write_timeout) || -1).to_f
+      @connect_timeout            = (params.delete(:connect_timeout) || (@read_timeout/2)).to_f
+      buffered                    = params.delete(:buffered)
+      @buffered                   = buffered.nil? ? true : buffered
+      @connect_retry_count        = params.delete(:connect_retry_count) || 10
+      @retry_count                = params.delete(:retry_count) || 3
+      @connect_retry_interval     = (params.delete(:connect_retry_interval) || 0.5).to_f
+      @on_connect                 = params.delete(:on_connect)
+      @write_timeout              = (params[:write_timeout] || 60.0).to_f
+      @ssl                        = params.delete(:ssl)
+      @ssl                        = false if ssl.nil?
+      @ssl_connect_timeout        = (params.delete(:ssl_connect_timeout) || 30.0).to_f
+      @ssl_context_params         = (params.delete(:ssl_context_params) || {})
+      @ssl_connect_retry_count    = params.delete(:ssl_connect_retry_count) || 10
+      @ssl_connect_retry_interval = (params.delete(:ssl_connect_retry_interval) || 0.5).to_f
```

Comment:
The thing I don't like about the mongo implementation is you should always call `set_params`, I think, even if it's `set_params({})` because it sets a lot of defaults.

I think limiting the set of options is fine (but there are times when you have to downgrade SSL or something so all the other random options are helpful to have).

"Most" users shouldn't have to set anything, so it should just be:

```ruby
Net::TCPClient.connect(
  servers:         ['server:3300', 'server2:3300'],
  server_selector: :random,
  ssl:             true
)
```